man1/ekaviel.1
.TH ekaviel 1 v1.53b Ekaviel
.PD
.SH NAME
ekaviel \- music module player
.SH SYNOPSIS
\fBekaviel\fR [\fIoptions\fP] [\fIfilename\fP] [...]
.SH DESCRIPTION
\fIEkaviel\fP plays music files (aka 'modules') via the computer's
audio channels. Modules are \fInot\fP just samples, but files that
contain a digital form of musical score, and the instruments
(often in the form of samples) built in. Such modules can be
produced with popular editors, such as SoundTracker and MED/
OctaMED
.br
For further details, see the DETAILS section
.SH OPTIONS
.TP 8
.B -bleed\fP <0-255>
Simulate being in a room on headphones.
.br
0 is pure stereo,
.br
128 is 100% bleed (ie. mono)
.br
255 is pure stereo, left and right speakers swapped
.TP 8
.B -gui/-nogui\fP
Run ekaviel with/without gui (override default)
.br
.TP 8
.B -interact\fP <n>
Set the interactivity level:
.br
0 is off (no keyboard input)
.br
1 is minimal keyboard input checking
.br
2 is most frequent checking of input; because of the
cpu-intensive nature, \fB-outrate 8000\fP is implicit when
\fB-interact 2\fP is set.
.TP 8
.B -limit\fP
Use a lookup table for extra speed. Nonstandard note
frequencies are not allowed. Is the default for Irix
4 machines. See also \fB-unlimit\fP
.TP 8
.B -loop\fP
Makes \fIekaviel\fP loop round the songs in the song list,
until the ESCape key is pressed
.br
** NOTE **: The ESCape key is ignored if \fB\-interact 0\fP
has been specified
.TP 8
.B -outrate\fP <speed>
Set the output rate in hertz (where hardware allows)
Large values (typically 44.1K, ie. CD quality) imply
greater CPU demand, which cannot necessarily be met;
if not met, this results in "stuttered" output.
.br
Refer to the audio manual pages for the hertz values
that the hardware can accept. Default value ie 22050
unless full interaction has been requested
.TP 8
.B -repeat\fP <times>
Play (without reloading) the next named song a given
number of times. Zero means zero, not infinity.
.TP 8
.B -seq\fP <list-posn>
Start the song from the given offset in the sequence
list (the default value is zero, which is always the
position number for the start)
.TP 8
.B -shuffle\fP
Random play (initially generates a seed with \fItime\fP(2)
and uses \fIrandom\fP(3B) to make random-sized jumps round
a circular linked list of unplayed modules)
.TP 8
.B -stereo\fP | \fB\-mono\fP
Put the audio hardware in stereo or mono output mode
(where available). The \fB-bleed\fP setting has no meaning
where \fB-mono\fP has been specified.
.TP 8
.B -sync
Enable synchronisation of audio and display
.TP 8
.B -transpose\fP <n>
Set the pitch transpose. Overrides transpose in song
Sadly, -ve numbers look like bad options...
.br
.TP 8
.B -unlimit\fP
Forces \fIEkaviel\fP to calculate all frequency values, at
the expense of processor time. Removes limits on the
values allowed that are normally imposed by software
See also \fB-limit\fP
.TP 8
.B -verbose\fP <n>
Set verbosity level. Is a bitwise sum, of:
.br
1: 0=Suppress/1=Display "playback" text
.br
2: 0=Suppress/1=Display "status" text
.br
4: 0=Suppress/1=Display "warning" text
.TP 8
.B -volume\fP <0-63>
Specify the volume at which to play back the module.
The range of \fB-volume\fP is not connected with the audio
hardware. Use of \fB-volume\fP overrides any master volume
setting present in the music files.
.br
It is a \fBfeature\fP of \fIEkaviel\fP that volumes of 64 and up
can be specified at run-time, thus allowing the user
to amplify quiet modules. Volume levels greater than
the recommended range will ordinarily produce output
that is saturated (akin to having corrupt samples).
.SH CONTROLS
During playback, the following controls are available:
.TP 8
.B ^N
Jump to next song in list
.TP 8
.B ^V
Freeze/unfreeze scrolling display
.TP 8
.B ESCape
Quit the player
.TP 8
.B <
Rewind to previous block in song
.TP 8
.B >
Fast forward to next block in song
.SH MENUS
The main menu can be recognised by the string '-EKAVIEL-' in
the top left hand corner of the screen. Activating sub-menus
is achieved by pressing a combination of the control key and
some other letter. For the '\fB^T\fPune' menu, this is CTRL-T.
.br
For the other menus, you will see something like '[TUNE]' in
the top left hand corner of the screen. The first letters of
the options (without CTRL) will activate them. At this point
the usual effects of pressing the keys alone are disabled.
.SH DETAILS
\fIEkaviel\fP runs in a curses-driven window, and has a basic menu
and keyboard-driven interface once started.
.SH CAPABILITIES
\fIEkaviel\fP is the successor to modplay, another player, and the
following additional facilities are provided:
.br
o SGI audio is handled in a less dodgy fashion
.br
o \fIEkaviel\fP provides additional Soundtracker effect support
.br
o Improved MMD0 support
.br
o Additionally handles MMD1 format (OctaMED v3+)
.br
o Supports loading of many popular compression formats
.SH CAVEATS
o \fIEkaviel\fP can't handle non-sample instruments (like modplay)
o \fIEkaviel\fP's MMD2 support is buggy
.br
Only displays four channels (not important yet)
.br
The curses screen driver library (used for the "ekgui" front
end) may allow the stopping of ekaviel with ctrl-break. This
is a feature of the curses library on some platforms, not of
the ekaviel source code. In theory, a ctrl-break event could
stop ekaviel prior to its deletion of a temporary file...
.SH FILES
~wills/bin/SG/ekaviel (Irix4 version)
.br
~wills/misc/mods/*.mod.lha
.br
~wills/misc/mods/*.mmd.lha
.SH ENVIRONMENT
.TP 8
.B $MODPATH\fP
Where to locate modules not in the current directory
.SH COMPILE-TIME
.TP 8
.B -DDEFAULT_OUTRATE=<value>
Default \fBoutrate\fP level for interactivity level 2
.TP 8
.B -DHAVE_EKGUI
Use the curses sceen handling libraries (rather than
stdin and stdout)
.TP 8
.B -DSLOWCPU
Use -limit by default, rather than -unlimit
.TP 8
.B -DSLOW_DISPLAY
Ekaviel normally maintains a buffer for each note of
the song, and sprintf() is used at load time to make
the text representation. If SLOW_DISPLAY is defined,
the sprintf() calls are mid-playback instead and the
memory requirement is less (but CPU demand greater).
.SH BUGS
That blond fella? That's Henry off of Neighbours, that is!
.SH But seriously:
Please mail me a bug-report if you find any. I suspect there
may be quite a few, so happy hunting! No reports for missing
effects, I know about those.
.SH AUTHOR
William "Wills" Towle, wills@scs.leeds.ac.uk


src/ekaviel.c
#ident "$ ekaviel v1.53b 16/02/96 Wm.Towle $"

/*****
 *	 Easther, Ke'lle', Vernie, Louise:
 *	 	" No matter what they say,
 *	 	No words I know can keep me away from your love;
 *	 	They say I'm crazy, when it comes to you. "
 *	 You will forever remain a great influence to me.
 *	 May you live long and prosper. Stay gorgeous,
 *	 						Wills.
 *****/

#include <math.h>	/* srandom() */
#include <stdlib.h>	/* atoi(), exit() */
#include <string.h>
#include <time.h>	/* time() in random number generation */
#include <unistd.h>	/* unlink() */
#include "ekaviel.h"

#include "argcv.h"
#include "ekmalloc.h"
#include "eklist.h"
#ifdef HAVE_EKGUI
#include "ekgui.h"
#endif
#include "ekio.h"
#include "eksong.h"
#include "loadbase.h"	/* find_file(), compression(), uncompress() */
#include "player.h"	/* play_song() */
#include "vchannel.h"	/* period_l() / period_u() */

typedef enum ekParseState {
	parseState_initial, parseState_midoptions,
	parseState_postsong, parseState_error
	} EkParseState;

EkErrFlag ekplaylist_new(EkPlayList **listPtr)
{
	if (listPtr == NULL)
	{
#ifdef DEVEL
	    fprintf(stderr, "ekplaylist_new() called %s\n",
				"with NULL pointer");
#endif
	    return EK_ERR_MALLOC_FAILURE;
	}
	else
	{
	  EkPlayList	*list= (EkPlayList *)
				ALLOCATE_MEMORY(1, EkPlayList,
					"ekaviel::ekplaylist_new()");

	    if ( (*listPtr= list) == NULL )
		return EK_ERR_MALLOC_FAILURE;
	    else
		return EK_ERR_NONE;
	}
}

EkErrFlag ekplaylist_construct(EkPlayList *list, EkPrefs *prefs)
{
	if (list == NULL)
	{
#ifdef DEVEL
	    fprintf(stderr, "ekplaylist_construct() called %s\n",
				"with NULL pointer");
#endif
	    return EK_ERR_MALLOC_FAILURE;
	}
	else
	{
	    EKPLAYLIST_FIELD(*list, filename)= NULL;
	    return eksetup_construct(
			&EKPLAYLIST_FIELD(*list, settings), prefs);
	}
}

EkErrFlag ekplaylist_destroy(EkPlayList *list)
{
	if (list == NULL)
	{
#ifdef DEVEL
	    fprintf(stderr, "ekplaylist_destroy() called %s\n",
				"with NULL pointer");
#endif
	    return EK_ERR_MALLOC_FAILURE;
	}
	else
	    return eksetup_destroy(&EKPLAYLIST_FIELD(*list, settings));
}

EkErrFlag ekplaylist_delete(EkPlayList **listPtr)
{
	if (listPtr == NULL)
	{
#ifdef DEVEL
	    fprintf(stderr, "ekplaylist_delete() called %s\n",
				"with NULL pointer");
#endif
	    return EK_ERR_MALLOC_FAILURE;
	}
	else
	{
	    FREE_MEMORY(*listPtr, "ekaviel::ekplaylist_delete()");
	    *listPtr= NULL;
	    return EK_ERR_NONE;
	}
}

#ifndef DEVEL
static
#endif
EkErrFlag playlist_clear(EkList *plist)
{
  EkErrFlag	error= EK_ERR_NONE;

	while ( eklist_next(plist) && (error == EK_ERR_NONE) )
	{
	  EkPlayList	*listent= (EkPlayList *)eklist_current(plist);

	    if ( (error= ekplaylist_destroy(listent)) == EK_ERR_NONE )
		error= ekplaylist_delete(&listent);
	    if (error == EK_ERR_NONE)
		error= eklist_unlink(plist);
	}

  return error;
}

#ifndef DEVEL
static
#endif
EkErrFlag play_playlist(EkList *unplayed,
#ifdef HAVE_EKGUI
			Flag ekgui,
#endif
			Flag shuffle, Flag loopOption)
{
  EkErrFlag	error;
  EkList	played;
  int		randmax= 0;
  EkPlayList	*lastSong;
  Flag		firstSong;
  EkSong	song;

#ifdef HAVE_EKGUI
	error= io_init(ekgui? gui_init : NULL);
#else
	error= io_init();
#endif
	if ( (error != EK_ERR_NONE)
		|| ((error= eklist_construct(&played)) != EK_ERR_NONE) )
	{
	    fprintf(stderr, "Initialise failure -%s\n",
				EKERRFLAG_STRING_VALUE(error));
	    return error;
	}

	if (shuffle)
	{
	    while ( eklist_next(unplayed) )
		randmax++;
	    eklist_set_circular(unplayed, FL_TRUE);
	    srandom((unsigned int) time(NULL));
	}

	firstSong= FL_TRUE;
	lastSong= NULL;	/* either last song played or load-attempted */
	do
	{
	  Flag quitNow= FL_FALSE;

	    while ( eklist_next(unplayed)
			&& (quitNow == FL_FALSE)
			&& (error == EK_ERR_NONE) )
	    {
	      EkPlayList	*listent;
	      int		command;
	      char		*tempfile;
	      Flag		doLoad;

		if (shuffle)
		{
		  int i;
		    for (i=0; i<(random() % randmax); i++)
			eklist_next(unplayed);
		}
		listent= (EkPlayList *)eklist_current(unplayed);
		doLoad= ( (lastSong == NULL)
			|| (strcmp(
				EKPLAYLIST_FIELD(*listent, filename),
				EKPLAYLIST_FIELD(*lastSong, filename))
				!= 0) )
			? FL_TRUE : FL_FALSE;

		error= EK_ERR_NONE;	/* hmmm, if doLoad "corrupt"? */
		tempfile= NULL;
		if (doLoad)
		{
		  int	compressID;
		    if ((compressID= compression(
			EKPLAYLIST_FIELD(*listent, filename))) != -1)
		    {
			stat2("Uncompressing",
				EKPLAYLIST_FIELD(*listent, filename));
			tempfile= uncompress(
				EKPLAYLIST_FIELD(*listent, filename),
				compressID);
		    }

		    if (firstSong)
			firstSong= FL_FALSE;
		    else
			error= eksong_destroy(&song);

		    error= eksong_construct(&song);
		}

		if (error == EK_ERR_NONE)
		{
		  EkErrFlag serror= EK_ERR_NONE;

		    if (tempfile)
		    {
			serror= eksong_load_bestguess(&song,
							tempfile);
			stat1("Cleaning temporary files...");
			unlink(tempfile);
			FREE_MEMORY(tempfile, "ekaviel::play_playlist()");
		    }
		    else if (doLoad)
			serror= eksong_load_bestguess(&song,
				EKPLAYLIST_FIELD(*listent, filename));

		    lastSong= listent;

		    if (serror != EK_ERR_NONE)
		    {
			warn3(EKPLAYLIST_FIELD(*listent, filename),
				"is unplayable -",
				EKERRFLAG_STRING_VALUE(serror));
		    }
		    else
		    {
		      int repeatTimes= EKSETUP_FIELD(
					EKPLAYLIST_FIELD(
						*listent, settings),
						repeatTimes);
					
			if ((eksong_instrs_playable(&song) == 0)
				&& (eksong_instrs_unplayable(&song) > 0)
				)
			{
			    warn2("All instruments unplayable in",
				EKPLAYLIST_FIELD(*listent, filename));
			}

			while ( !quitNow && (repeatTimes > 0) )
			{
			    stat2("Playing",
				EKPLAYLIST_FIELD(*listent, filename));

			    error= play_song(&song, &EKPLAYLIST_FIELD(
						*listent, settings),
						&command);
			    if ( error != EK_ERR_NONE )
			    {
				quitNow= FL_TRUE;
				stat2("Playback error:",
					EKERRFLAG_STRING_VALUE(error));
			    }
			    else if (command != EKVL_C_QUIT)
				repeatTimes--;
			    else
				quitNow= FL_TRUE;
			}
		    }
		}
		else if (tempfile)
		{	/* Clean up, no memory to do anything. */
		    stat1("Cleaning temporary files...");
		    unlink(tempfile);
		    FREE_MEMORY(tempfile, "ekaviel::play_playlist()");
		}

		eklist_add(&played, listent);
		eklist_unlink(unplayed);
	    }

	    if (loopOption)
	    {
		/* put "played" songs back on "unplayed" list */
		while (eklist_next(&played))
		{
		    eklist_add(unplayed, eklist_current(&played));
		    eklist_unlink(&played);
		}

		if (quitNow) loopOption= FL_FALSE;
	    }
	}
	while (loopOption);

	error= eksong_destroy(&song);


	/* clear "unplayed" in case quit with songs yet to play */
	if ((error= playlist_clear(unplayed)) != EK_ERR_NONE)
	{
	    fprintf(stderr, "Terminate failure -%s\n",
				EKERRFLAG_STRING_VALUE(error));
	    return error;
	}

	/* clear "played" in case songs as yet unplayed */
	if ( ((error= playlist_clear(&played)) != EK_ERR_NONE)
		|| ((error= eklist_destroy(&played)) != EK_ERR_NONE) )
	{
	    fprintf(stderr, "Terminate failure -%s\n",
				EKERRFLAG_STRING_VALUE(error));
	    return error;
	}

#ifdef HAVE_EKGUI
	return io_restore(ekgui? gui_restore : NULL);
#else
	return io_restore();
#endif
}

int main(int argc, const char **argv)
{
  Flag		done;
  const	char	*argText, *parText;
  EkErrFlag	error;
  EkParseState	pState;
  EkList	playList;
  EkPrefs	defPrefs;

	if ( ((error= eklist_construct(&playList)) != EK_ERR_NONE)
		|| ((error= ekprefs_construct(&defPrefs))
						!= EK_ERR_NONE) )
	{
	    fprintf(stderr, "Initialise failure: %s\n",
					EKERRFLAG_STRING_VALUE(error));
	    exit(1);
	}

	initArg(argc, argv);

	done= FL_FALSE;
	pState= parseState_initial;
	while ( done == FL_FALSE )
	{
	  EkPlayList	*listent= NULL;

	    if ( ((error= ekplaylist_new(&listent)) != EK_ERR_NONE)
		|| ((error= ekplaylist_construct(listent, &defPrefs))
						!= EK_ERR_NONE) )
	    {
		fprintf(stderr, "Initialise failure -%s\n",
					EKERRFLAG_STRING_VALUE(error));
		pState= parseState_error; done= FL_TRUE;
	    }

	    EKPREFS_FIELD(defPrefs, interact)=
			EKSETUP_FIELD( EKPLAYLIST_FIELD(
				*listent, settings), interact)= 2;
	    EKPREFS_FIELD(defPrefs, verbose)=
			EKSETUP_FIELD( EKPLAYLIST_FIELD(
				*listent, settings), verbose)= 7;

	    while (nextOpt(&argText, &parText))
	    {
	      Flag	missing= FL_FALSE;

		if (strcmp(argText, "bleed") == 0)
		{
		    if (parText)
			defPrefs.bleed= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					bleed)= atoi(parText);
		    else
			missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "gui") == 0)
		{
#ifdef HAVE_EKGUI
		    defPrefs.hasgui= FL_TRUE;
#endif
		    if (parText) rewindArg();
		}
		else if (strcmp(argText, "nogui") == 0)
		{
#ifdef HAVE_EKGUI
		    defPrefs.hasgui= FL_FALSE;
#endif
		    if (parText) rewindArg();
		}

		else if (strcmp(argText, "interact") == 0)
		{
		    if (parText)
		    {
			if ((defPrefs.interact= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					interact)= atoi(parText)) == 2)
			{
#ifdef I2RATE
			    defPrefs.outrate= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					outrate)= I2RATE;
#else
			    defPrefs.outrate= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					outrate)= 8000;
#endif
			}
		    }
		    else
			missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "limit") == 0)
		{
		    defPrefs.periodfn= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					periodfn)= period_l;
		    if (parText) rewindArg();
		    pState= parseState_midoptions;
		}
		else if (strcmp(argText, "unlimit") == 0)
		{
		    defPrefs.periodfn= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					periodfn)= period_u;
		    if (parText) rewindArg();
		}

		else if (strcmp(argText, "loop") == 0)
		{
		    if (pState != parseState_initial)
		    {
			fprintf(stderr,
				"Expected '-loop' before playlist\n");
			pState= parseState_error; done= FL_TRUE;
		    }
		    else
		    {
			defPrefs.loop= FL_TRUE;
			if (parText) rewindArg();
		    }
		}

		else if (strcmp(argText, "outrate") == 0)
		{
		    if (parText)
			defPrefs.outrate= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					outrate)= atoi(parText);
		    else missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "repeat") == 0)
		{
		    if (parText)
			EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					repeatTimes)= atoi(parText);
		    else missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "seq") == 0)
		{
		    if (parText)
			EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					seqSkip)= atoi(parText);
		    else missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "shuffle") == 0)
		{
		    if (pState != parseState_initial)
		    {
			fprintf(stderr, "Expected '-shuffle' before playlist\n");
			pState= parseState_error; done= FL_TRUE;
		    }
		    else
		    {
			defPrefs.shuffle= FL_TRUE;
			if (parText) rewindArg();
		    }
		}

		else if (strcmp(argText, "stereo") == 0)
		{
		    defPrefs.stereo= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					stereo)= FL_TRUE;
		    if (parText) rewindArg();
		    pState= parseState_midoptions;
		}
		else if (strcmp(argText, "mono") == 0)
		{
		    defPrefs.stereo= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					stereo)= FL_FALSE;
		    if (parText) rewindArg();
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "sync") == 0)
		{
		    defPrefs.synchro= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					synchro)= FL_TRUE;
		    if (parText) rewindArg();
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "transpose") == 0)
		{
		    if (parText)
			EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					transpose)= atoi(parText);
		    else missing= FL_TRUE;
		    pState= parseState_midoptions;
		}
		else if (strcmp(argText, "downtranspose") == 0)
		{
		    if (parText)
			EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					transpose)= -atoi(parText);
		    else missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "verbose") == 0)
		{
		    if (parText)
			defPrefs.verbose= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					verbose)= atoi(parText);
		    else
			missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else if (strcmp(argText, "volume") == 0)
		{
		    if (parText)
			defPrefs.volume= EKSETUP_FIELD(
				EKPLAYLIST_FIELD(*listent, settings),
					volume)= atoi(parText);
		    else
			missing= FL_TRUE;
		    pState= parseState_midoptions;
		}

		else
		{
		    fprintf(stderr, "Bad argument '%s'\n", argText);
		    pState= parseState_error; done= FL_TRUE;
		}

		if (missing)
		{
		    fprintf(stderr, "Missing argument to option %s\n",
					argText);
		    pState= parseState_error; done= FL_TRUE;
		}
	    }

	    if (! nextArg(&argText) )
	    {
		if ( (pState == parseState_initial)
			|| (pState == parseState_midoptions) )
		{
		    fprintf(stderr, "Expected song file\n");
		    pState= parseState_error; done= FL_TRUE;
		}

		done= FL_TRUE;
	    }
	    else
	    {
		if ( (EKPLAYLIST_FIELD(*listent, filename)=
					find_file(argText)) == NULL )
		{
		    fprintf(stderr, "Couldn't open %s\n", argText);
		    if ( ((error= ekplaylist_destroy(listent))
						!= EK_ERR_NONE)
			|| ((error= ekplaylist_delete(&listent))
						!= EK_ERR_NONE) )
		    {
			fprintf(stderr, "Initialise failure -%s\n",
					EKERRFLAG_STRING_VALUE(error));
			pState= parseState_error; done= FL_TRUE;
		    }
		}
		else
		{
		    eklist_add(&playList, listent);
		    pState= parseState_postsong;
		}
	    }
	}

	switch (pState)
	{
	case parseState_postsong:
	    if (play_playlist(&playList,
#ifdef HAVE_EKGUI
				defPrefs.hasgui,
#endif
				defPrefs.shuffle, defPrefs.loop)
			!= EK_ERR_NONE)
	    {
		exit(1);	/* some failure (already reported) */
	    }
	    break;
	case parseState_error:
	    fprintf(stderr, "\n");
	default:
	    fprintf(stderr,
		"Usage:\n\t%s [options] < [prefs] song >[...]\n\n",
			argv[0]);
	    fprintf(stderr, "%s\n\t%s\t%s\n\t%s\t%s\n",
			"Options:",
			"-loop\t", "Cycle infinitely through songs",
			"-shuffle", "Play songs in random order");
	    fprintf(stderr, "%s\n\t%s\t%s\n\t%s\t%s\n\t%s\t%s\n",
			"Prefs:",
			"-bleed <n>", "balance control",
			"-gui/-nogui", "with/without 'cute' front-end",
			"-interact <n>", "keyboard sensitivity");
	    fprintf(stderr, "\t%s\t%s\n\t%s\t%s\n\t%s\t%s\n",
			"-limit/-unlimit", "note accuracy",
			"-outrate", "output quality",
			"-repeat <n>", "individual song repetition");
	    fprintf(stderr, "\t%s\t%s\n\t%s\t%s\n\t%s\t%s\n",
			"-seq <n>", "jump-point into song",
			"-stereo|-mono", "playback in 1/2 channels",
			"-sync\t", "display control");
	    fprintf(stderr, "\t%s\t%s\n\t%s\t%s\n\t%s\t%s\n",
			"-transpose|-downtranspose <n>",
					"output pitch control",
			"-verbose <n>", "output level control",
			"-volume <n>", "volume level control");
	}

	if ( ((error= ekprefs_destroy(&defPrefs)) != EK_ERR_NONE)
		|| ((error= playlist_clear(&playList)) != EK_ERR_NONE)
		|| ((error= eklist_destroy(&playList)) != EK_ERR_NONE) )
	{
	    fprintf(stderr, "Terminate failure: %s\n",
					EKERRFLAG_STRING_VALUE(error));
	    exit(1);
	}

  return 0;
}
